File: src/dfsAndbfs/NofCElements.java
package dfsAndbfs;

import java.io.*;
import java.util.Stack;

public class NofCElements {
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] nm = br.readLine().split(" ");
        int n = Integer.parseInt(nm[0]), m = Integer.parseInt(nm[1]);
        int[][] graph = new int[n + 1][n + 1];
        boolean[] visited = new boolean[n + 1];

        Stack<Integer> stack = new Stack<>();
        int count = 0;

        for (int i = 0; i < m; i++) {
            String[] ab = br.readLine().split(" ");
            int a = Integer.parseInt(ab[0]), b = Integer.parseInt(ab[1]);

            graph[a][b] = 1;
            graph[b][a] = 1;
        }

        while(includes(visited)){
            int node = findIndex(visited);
            count++;
            stack.push(node);
            visited[node] = true;

            while(!stack.empty()){
                int cur = stack.pop();

                for (int i = 1; i < n + 1; i++) {
                    if(graph[cur][i] == 1 && !visited[i]){
                        stack.push(i);
                        visited[i] = true;
                    }
                }
            }
        }

        bw.write(String.valueOf(count));

        bw.flush();
        bw.close();
    }

    static boolean includes(boolean[] visited){
        for (int i = 1; i < visited.length; i++) {
            if(!visited[i]) return true;
        }

        return false;
    }

    static int findIndex(boolean[] visited){
        for(int i = 1; i < visited.length; i++){
            if(!visited[i]) return i;
        }

        return -1;
    }
}

File: JS/dfs&bfs/NofCElements.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().trim().split("\n")
    : `6 8
1 2
2 5
5 1
3 4
4 6
5 4
2 4
2 3`.split("\n");

const [n, m] = input.shift().split(" ").map(Number);
const graph = Array.from(Array(n + 1), () => Array(n + 1).fill(0));
const visited = Array(n + 1).fill(false);
const stack = [];
let count = 0;

input.forEach((v) => {
  const [a, b] = v.split(" ").map(Number);
  graph[a][b] = 1;
  graph[b][a] = 1;
});

while (visited.slice(1).includes(false)) {
  const node = visited.findIndex((v, i) => i > 0 && v === false);
  count++;
  stack.push(node);
  visited[node] = true;

  while (stack.length > 0) {
    const cur = stack.pop();

    for (let i = 1; i <= n; i++) {
      if (graph[cur][i] === 1 && !visited[i]) {
        stack.push(i);
        visited[i] = true;
      }
    }
  }
}

console.log(count);
